RHDEVDOCS-891: Reviewed the Workspace Data Model section for language… (#608)
codenvy/che-docs: src/main/pages/workspace-admin/workspace-data-model.adoc

title: "Workspace Data Model"
keywords: workspace, runtime, recipe, docker, yaml, Dockerfile, docker, kubernetes, container, pod
tags: [workspace, runtime, docker, kubernetes, dev-docs]
sidebar: che_6_docs
permalink: workspace-data-model.html
folder: workspace-admin
---

The following table lists the data types and their description.

[width="100%",cols="50%,50%",options="header",]
|===
|Data Types |Description
|environments: Map<String, getEnvironments> 
|Workspace environment variables. A workspace can have multiple environment variables.

|defaultEnv: STRING
|A workspace must have a default environment.

|projects: []
|Projects associated with a workspace.

|commands: []
|Commands associated with a workspace.

|name: STRING
|Workspace name that has to be unique in a namespace.

|links: []
|-
|===

[id="environment"]
== Environment recipes

image::workspaces/ws_data_model.png[]

> Environment recipe can also have content instead of location, if the link:recipes.html[recipe] type is `dockerfile`, `compose` or `openshift`:

Reviewer: serial comma
```suggestion
For link:recipes.html[recipe] types of `dockerfile`, `compose`, or `openshift`, content, not location, is specified.
```

image::workspaces/ws_projects.png[]

The project object structure has the `source.location` and `source.type` parameters. There are two importer types: `git` and `zip`. New location types can be provided by custom plugins, such as `svn`.

> If a project is misconfigured or missing sources, it is marked with an error code and a message explaining the error. In the example above, the project does not have errors and mixins.

Reviewer: if
```suggestion
Incorrectly configured projects or projects missing sources are marked with error codes and messages explaining the error. In the example above, the project does not have errors and mixins.
```

[id="commands"]
== Commands

Commands can be both tied to a workspace and an individual project. In the example below, a command is saved to workspace configuration.

image::workspaces/ws_commands.png[]

The followling image shows ways to save commands in the project configuration.

image::workspaces/project_commands.png[]

[id="runtime"]
== Runtime

> Runtime object is created when a workspace is in a running state. Runtime returns server URLs, internal or external, depending on the server configuration. Interested clients, like User Dashboard and IDE, use these URLs.

Reviewer: ```suggestion
A runtime object is created when a workspace is in a running state. Runtime returns server URLs, internal or external, depending on the server configuration. Interested clients, like the *User Dashboard* and the IDE, use these URLs.
```
